code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/utility/FileUtils.kt

```diff
@@ -0,0 +1,199 @@
+     * Extracts the zip file to an output directory.
+     *
+     * @param zipFile the zip file that needs to be extracted
+     * @param outputDirectoryPath the destination for the extracted [zipFile] contents
+     * @return true if the zip file has been successfully extracted
+     *         false otherwise
+     */
+    @JvmStatic
+    fun extractFromZip(zipFile: File, outputDirectoryPath: String): Boolean {
+        val folder = File(outputDirectoryPath)
+        if (!folder.exists() && !folder.mkdir()) {
+            MobileCore.log(
+                LoggingMode.WARNING,
+                TAG,
+                "Could not create the output directory $outputDirectoryPath"
+            )
+            return false
+        }
+
+        var entryProcessedSuccessfully = true
+        try {
+            ZipInputStream(FileInputStream(zipFile)).use { zipInputStream ->
+                // get the zipped file list entry
+                var ze = zipInputStream.nextEntry
+                val outputFolderCanonicalPath = folder.canonicalPath
+                if (ze == null) {
+                    // Invalid zip file!
+                    entryProcessedSuccessfully = false
+                    MobileCore.log(LoggingMode.WARNING, TAG, "Zip file was invalid")
+                }
+
+                while (ze != null && entryProcessedSuccessfully) {
+                    val fileName = ze.name
+                    val newZipEntryFile = File(outputDirectoryPath + File.separator + fileName)
+                    if (!newZipEntryFile.canonicalPath.startsWith(outputFolderCanonicalPath)) {
+                        MobileCore.log(
+                            LoggingMode.ERROR,
+                            TAG,
+                            "The zip file contained an invalid path. Verify that your zip file is formatted correctly and has not been tampered with."
+                        )
+                        return false
+                    }
+                    entryProcessedSuccessfully = if (ze.isDirectory) {
+                        // handle directory
+                        (newZipEntryFile.exists() || newZipEntryFile.mkdirs())
+                    } else {
+                        // handle file
+                        val parentFolder = newZipEntryFile.parentFile
+                        if (parentFolder != null && (parentFolder.exists() || parentFolder.mkdirs())) {
+                            readInputStreamIntoFile(newZipEntryFile, zipInputStream, false)
+                        } else {
+                            MobileCore.log(
+                                LoggingMode.WARNING,
+                                TAG,
+                                "Could not extract the file ${newZipEntryFile.absolutePath}"
+                            )
+                            return false
+                        }
+                    }
+                    ze = zipInputStream.nextEntry
+                }
+                zipInputStream.closeEntry()
```

Comment:
Should we call `closeEntry()` before moving to the next entry?

Reply:
`.nextEntry` will check and close the current open entry if any. But agree that it is safer to be explicit. Added `closeEntry()` before moving to the next extry.